lib/teamsnap.rb

```diff
@@ -131,18 +135,51 @@ def init(opts = {})
       apply_endpoints(self, collection) && true
     end
 
-    def run(via, href, args = {})
+    def run(via, href, args = {}, opts = {})
       resp = client.send(via, href, args)
 
       if resp.status == 200
-        Oj.load(resp.body)
-          .fetch(:collection)
+        collection = Oj.load(resp.body).fetch(:collection)
+        if opts.include?(:backup_cache_file)
+          TeamSnap.write_backup_file(opts[:backup_cache_file], collection)
+        end
+        collection
       else
-        error_message = parse_error(resp)
-        raise TeamSnap::Error.new(error_message)
+        if TeamSnap.backup_file_exists?(opts)
+          warn("Connection to API failed.. using backup cache file to initialize endpoints")
+          Oj.load(IO.read(opts[:backup_cache_file]))
+        else
+          error_message = parse_error(resp)
+          raise TeamSnap::Error.new(error_message)
+        end
       end
     end
 
+    def backup_file(backup_cache)
```

Comment:
I'd name this something like `backup_file_for` to show you're returning something vs backing up a file.